File: Python/107-binary-tree-level-order-traversal-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if root == None:
            return []
        res = []
        queue = [root]
        while queue != []:
            row, size = [], len(queue)
            while size != 0:
                node = queue.pop(0)
                row.append(node.val)
                if node.left != None:
                    queue.append(node.left)
                if node.right != None:
                    queue.append(node.right)
                size -= 1
            res.append(row)
        res.reverse()
        return res

File: Go/107-binary-tree-level-order-traversal-ii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func reverse(nums [][]int) [][]int {
    start, end := 0, len(nums) - 1
    for start < end {
        nums[start], nums[end] = nums[end], nums[start]
        start++
        end--
    }
    return nums
}

func levelOrderBottom(root *TreeNode) [][]int {
    var result [][]int
    if root == nil {
        return result
    }
    queue := []*TreeNode{root} //队列
    for len(queue) != 0 {
        curSize := len(queue)  //当前层的结点数
        var curRow []int
        for curSize != 0 {
            curNode := queue[0] //取队头元素并弹出
            queue = queue[1:]
            if curNode.Left != nil {
                queue = append(queue, curNode.Left)
            }
            if curNode.Right != nil {
                queue = append(queue, curNode.Right)
            }
            curRow = append(curRow, curNode.Val)
            curSize--
        }
        result = append(result, curRow)
    }
    result = reverse(result)
    return result
}